utils/format_doc.ml

```diff
@@ -0,0 +1,476 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Florian Angeletti, projet Cambium, Inria Paris             *)
+(*                                                                        *)
+(*   Copyright 2021 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+
+type box_type =
+  | H
+  | V
+  | HV
+  | HoV
+  | B
+
+type stag = Format.stag
+
+type element =
+  | Text of string
+  | With_size of int
+  | Open_box of { kind: box_type ; indent:int }
+  | Close_box
+  | Open_tag of Format.stag
+  | Close_tag
+  | Open_tbox
+  | Tab_break of { width : int; offset : int }
+  | Set_tab
+  | Close_tbox
+  | Simple_break of { spaces : int; indent: int }
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
+
+  | Deprecated of (Format.formatter -> unit)
+
+type doc = { rev:element list } [@@unboxed]
+
+type t = doc
+
+let empty : doc = { rev = [] }
+
+let to_list doc = List.rev doc.rev
+let add x doc = { rev = x :: doc.rev }
```

Comment:
I would find the parameter order `add doc elem` more natural than `add elem doc`, because the element is printed after the rest of the document, not before.